src/tbb/cgroup_info.h

```diff
@@ -209,31 +215,85 @@ class cgroup_info {
         if (!*paths_cache.v1_relative_path)
             cache_relative_path_for(cgroup_fd, paths_cache);
 
-        if (snprintf(dir, PATH_MAX, "%s/%s", mnt_dir, paths_cache.v1_relative_path) >= 0)
+        if (std::snprintf(dir, PATH_MAX, "%s/%s", mnt_dir, paths_cache.v1_relative_path) >= 0)
             try_read_cgroup_v1_num_cpus_from(dir, num_cpus);
         return num_cpus;
     }
 
+    static bool is_cpu_restriction_possible(std::FILE* cgroup_fd, cgroup_paths& relative_paths_cache) {
+        cache_relative_path_for(cgroup_fd, relative_paths_cache);
+
+        bool has_cgroup_v1_cpu_controller = *relative_paths_cache.v1_relative_path;
+        bool has_cgroup_v2 = *relative_paths_cache.v2_relative_path;
+        bool is_cgroup_v2_with_root_path =
+            relative_paths_cache.v2_relative_path[0] == '/' && !relative_paths_cache.v2_relative_path[1];
+        if (!has_cgroup_v2 && !has_cgroup_v1_cpu_controller) {
+            return false; // No CPU constraints found
+        } else if (has_cgroup_v1_cpu_controller || !is_cgroup_v2_with_root_path) {
+            return true;
+        }
+        __TBB_ASSERT(!has_cgroup_v1_cpu_controller && has_cgroup_v2 && is_cgroup_v2_with_root_path, nullptr);
+
+        // Usually PID 1 is the init process
+        unique_file_t init_process_cgroup_file(std::fopen("/proc/1/cgroup", "r"), &close_file);
+        cgroup_paths init_process_cgroup_cache{};
+        cache_relative_path_for(init_process_cgroup_file.get(), init_process_cgroup_cache);
+        if (*init_process_cgroup_cache.v2_relative_path) {
+            // If the init process cgroup path is "/init.scope", it means systemd is used
+            if (!std::strncmp(init_process_cgroup_cache.v2_relative_path, "/init.scope", 11)) {
+                return false;
+            }
+        }
```

Comment:
Why double check here? Does it give any speedup comparing to invocation of `std::strncmp` as in:
```suggestion
        // If the init process cgroup path is "/init.scope", it means systemd is used
        if (!std::strncmp(init_process_cgroup_cache.v2_relative_path, "/init.scope", 11)) {
            return false;
        }
```

Reply:
`v2_relative_path` is an array of chars, so it is always non-null here.